src/exec/noexec.rs

```diff
@@ -268,38 +309,7 @@ pub(crate) fn add_noexec_filter(command: &mut Command, file_closer: &mut FileClo
                 return Err(io::Error::last_os_error());
             }
 
-            let mut data = [0u8; 1];
-            let mut iov = iovec {
-                iov_base: &mut data as *mut [u8; 1] as *mut c_void,
-                iov_len: 1,
-            };
-
-            // SAFETY: msghdr can be zero-initialized
-            let mut msg: msghdr = zeroed();
-            msg.msg_name = ptr::null_mut();
-            msg.msg_namelen = 0;
-            msg.msg_iov = &mut iov;
-            msg.msg_iovlen = 1;
-
-            // SAFETY: SingleRightAnciliaryData can be zero-initialized.
-            let mut control: SingleRightAnciliaryData = zeroed();
-            // SAFETY: The buf field is valid when zero-initialized.
-            msg.msg_controllen = control.buf.len();
-            msg.msg_control = &mut control as *mut _ as *mut _;
-            // SAFETY: msg.msg_control is correctly initialized and this follows
-            // the contract of the various CMSG_* macros.
-            {
-                let cmsgp = CMSG_FIRSTHDR(&msg);
-                (*cmsgp).cmsg_level = SOL_SOCKET;
-                (*cmsgp).cmsg_type = SCM_RIGHTS;
-                (*cmsgp).cmsg_len = CMSG_LEN(size_of::<c_int>() as u32) as usize;
-                ptr::write(CMSG_DATA(cmsgp).cast::<c_int>(), notify_fd);
-            }
-
-            // SAFETY: A valid socket fd and a valid initialized msghdr are passed in.
-            if sendmsg(tx_fd.as_raw_fd(), &msg, 0) == -1 {
-                return Err(io::Error::last_os_error());
-            }
+            send_fd(&tx_fd, notify_fd)?;
```

Comment:
This can consume `tx_fd` just fine. It needs to be closed before the exec happens anyway. Actually it may well be that it currently doesn't get closed, which would leak the socket to the process sudo runs.

Reply:
The reason is that `tx_fd` is added to the `FileCloser`, by the way.

Reply:
`pre_exec` takes an `FnMut` closure in case you want to spawn the `Command` multiple times. Wrapping `tx_fd` and using `tx_fd.take().unwrap()` should work.

Reply:
It does.